scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala

```diff
@@ -223,38 +308,133 @@ class ScalametaParser(input: Input, dialect: Dialect) { parser =>
+                  curr.input,
+                  curr.dialect,
+                  curr.start,
+                  curr.end
+                )
+                parserTokenPositions += currPos
+              }
+              sepRegionsProcess = sepRegionsProcess.tail
+            }
+            if (!sepRegionsProcess.isEmpty) sepRegionsProcess = sepRegionsProcess.tail
+            sepRegionsProcess
+
           } else if (curr.is[RightBracket]) {
-            if (!sepRegions.isEmpty && sepRegions.head == ']') sepRegions.tail else sepRegions
+            if (!sepRegions.isEmpty && sepRegions.head.closing == RegionBracket) sepRegions.tail
+            else sepRegions
           } else if (curr.is[RightParen]) {
-            if (!sepRegions.isEmpty && sepRegions.head == ')') sepRegions.tail else sepRegions
+            var sepRegionsProcess = sepRegions
+            while (indentedRegion(sepRegionsProcess)) {
+              parserTokens += new Indentation.Outdent(
+                curr.input,
+                curr.dialect,
+                curr.start,
+                curr.end
+              )
+              parserTokenPositions += currPos
+              sepRegionsProcess = sepRegionsProcess.tail
+            }
+            if (!sepRegionsProcess.isEmpty && sepRegionsProcess.head.closing == RegionParen)
+              sepRegionsProcess.tail
+            else sepRegionsProcess
+          } else if (curr.is[LeftArrow]) {
+            if (!sepRegions.isEmpty && sepRegions.head.closing == RegionArrow) sepRegions.tail
+            else sepRegions
           } else if (curr.is[RightArrow]) {
-            if (!sepRegions.isEmpty && sepRegions.head == '\u21d2') sepRegions.tail else sepRegions
+            if (!sepRegions.isEmpty && sepRegions.head.closing == RegionArrow) sepRegions.tail
+            else sepRegions
           } else sepRegions // do nothing for other tokens
         }
-        loop(currPos, currPos + 1, sepRegions1)
+
+        parserTokens += curr
+        parserTokenPositions += currPos
+
+        @tailrec
+        def isAheadNewLine(pos: Int = currPos + 1): Boolean = {
+          if (pos >= scannerTokens.length) false
+          else if (scannerTokens(pos).is[LF]) true
+          else scannerTokens(pos).is[Trivia] && isAheadNewLine(pos + 1)
+        }
+
+        if (dialect.allowSignificantIndentation && (prev == null || prev.text != "end") && (curr
+            .is[KwYield] || curr.is[KwTry] || curr.is[KwCatch] || curr.is[KwFinally] || curr
+            .is[KwMatch] || curr.is[KwDo] || curr.is[KwFor] || curr.is[KwThen] || curr
+            .is[KwElse] || curr.is[Equals] || curr.is[KwWhile] ||
+          (curr.is[RightArrow] && (catchCaseIndent || !sepRegionsNew.headOption.exists(_.closing == RegionBrace)))) && isAheadNewLine()) {
+
+          if (curr.is[RightArrow] && sepRegions.headOption.exists(_.closing == RegionArrow)) {
+            // arrived at case <cond> '=>'
+            lastIndent = lastCaseIndent
```

Comment:
`lastIndent` seems to be only specified here, so it doesn't seem we need the separate variable.

Reply:
removed :)